app/delivery/send_to_providers.py

```diff
@@ -132,9 +132,11 @@ def send_email_to_provider(notification):
 
             if sending_method == "attach":
                 try:
-
+                    # Prevent URL patterns like file:// ftp:// that may lead to security vulnerabilities
+                    if not personalisation_data[key]["document"]["direct_file_url"].lower().startswith(("http", "s3")):
+                        raise ValueError("Invalid URL provided")
```

Comment:
My quick take on it is that we might want to validate this earlier if a `direct_file_url` is non secure. Applying the validation logic in that part of the code would raise an exception in a celery task. This error would not get back to the user. The celery task will get retried until the retries run out by the task and potentially throw celery errors alerts. 

Or do we want to have some defensive programming in here "just in case"?

It is certainly a good validation to have overall. Let me dig more into the code but it's my current understanding, I am not sure if we apply or not that validation already on the user front, we need to check.

Reply:
i'd leave this here just in-case. The test should flag this if we refactor.

If we can also track this upstream that's even better. Can users ever provide a URL themselves when setting notificaions? or will this always contain a s3 bucket url pointing to whatever file they uploaded? If a user can provide the URL directly then more work will have to be done upstream for sure.

Reply:
Hmm the `direct_file_url` entry might always target a S3 bucket if I read the code right, when the sending method of the notification is `attach`. But Notify does allow for linking (when sending method is `link`) and that would rather be on the `url` entry in the same dictionary. Realistically, then, I would expect the checked URL in this to be HTTPS and target S3 (to double check via some local debugging).

The `url` entry through provided by the user could be non HTTPS and could be a risk for people receiving the email. I don't [see anything in the documentation](https://documentation.notification.canada.ca/en/send.html#sending-a-file-by-email) that requires clients to provide a HTTPS link though. What do you think?

Reply:
I like it, that's even more well defined. I'll update this PR to restrict `url` to "https" and `direct_file_url` to "s3".

We can then update our docs to mention that https links are only allowed. This may cause problems for any GC departments still using http, but might also be an incentive to migrate to https